cirq-core/cirq/ops/controlled_gate.py

```diff
@@ -159,6 +160,12 @@ def _decompose_with_context_(
         self, qubits: Tuple['cirq.Qid', ...], context: Optional['cirq.DecompositionContext'] = None
     ) -> Union[None, NotImplementedType, 'cirq.OP_TREE']:
         control_qubits = list(qubits[: self.num_controls()])
+        # If the subgate is an EigenGate with non-zero phase, try to decompose it
+        # into a phase-free gate and a global phase gate.
+        if isinstance(self.sub_gate, eigen_gate.EigenGate) and self.sub_gate.global_shift != 0:
```

Comment:
Is there any way you can get rid of this condition? Like try decomposing the subgate first, then only enter the branch if the subgate has decomposed? A big goal is to reduce the amount of type checking needed here.

If this works, then the duplicate function call at the end of this function can be removed, and the function can just return NotImplemented if it gets to that point.